Feature/create arrears action diary entry

madetech-jeffpinkham opened the pull request:
API Endpoint for Creating action diary entries

madetech-jeffpinkham commented on LBHTenancyAPI/Services/CredentialsService.cs:
```diff
@@ -0,0 +1,21 @@
+    {
+        public string GetUhSourceSystem()
+        {
+            return "HackneyAPI";
```
Yep agreed

madetech-jeffpinkham replied:
I need to do the ENV piece we discussed on Tuesday, to accomodate that.

madetech-jeffpinkham replied:
I can do that as part of this task or as a separate task

madetech-jeffpinkham replied:
I'll do the ENV piece now 👍

StevenLeighton21 replied:
I think we should do that as part of this task, purely to avoid committing creds to the open source repo.

madetech-jeffpinkham replied:
YEAH! The repo is open source, isn't it.. Yep on it now.

StevenLeighton21 commented on LBHTenancyAPI/Gateways/ArrearsActionDiaryGateway.cs:
```diff
@@ -0,0 +1,24 @@
+{
+    public class ArrearsActionDiaryGateway : IArrearsActionDiaryGateway
+    {
+        private readonly IArrearsAgreementService _arrearsAgreementService;
```
Should it be called actionDiaryService? Or does the service cover both creating action diary entries and agreements?

madetech-jeffpinkham replied:
Yep agreed 👍

madetech-jeffpinkham commented:
I've moved the configuration into the ECS ENV variables, and used EnvironmentVariables in the configuration. I deleted the tests that tested the configuration for Credentials Service as it's configuration and I found it a nightmare to test.

zuren commented on LBHTenancyAPI.sln:
```diff
@@ -31,4 +33,10 @@ Global
+	GlobalSection(ExtensibilityGlobals) = postSolution
+		SolutionGuid = {7524071F-1952-4ABE-97E9-4E73D0F9BE47}
+	EndGlobalSection
 EndGlobal
```
Are these config changes actually necessary, or are they susceptible to churn every time a different dev pushes up?

madetech-jeffpinkham replied:
This is automatically done by Visual Studio, we could remove it but it would most likely get re-added by Visual Studio. It's subject to churn but Visual Studio usually manages it.

zuren commented on LBHTenancyAPI.v3.ncrunchsolution:
```diff
@@ -0,0 +1,6 @@
+    <AllowParallelTestExecution>True</AllowParallelTestExecution>
+    <SolutionConfigured>True</SolutionConfigured>
+  </Settings>
+</SolutionConfiguration>
```
Same here - possibly wise to gitignore? What's conventional?

madetech-jeffpinkham replied:
Yep we can gitignore this

zuren commented on LBHTenancyAPI/Connected Services/AgreementService/ConnectedService.json:
```diff
@@ -0,0 +1,24 @@
+  "ProviderId": "Microsoft.VisualStudio.ConnectedService.Wcf",
+  "Version": "15.0.20119.1312",
+  "ExtendedData": {
+    "Uri": "http://lbhuhwwebp01/civica-uh-gateway-dev/webservices/ArrearsAgreementService.svc",
```
Would consider this private, should be configured via the environment somehow IMO.

StevenLeighton21 replied:
That URI is on the live internet, no? It's accessible from outside prem? If so, I'd argue it's not private. If not, agree :)

zuren replied:
Nah, it's an internal host.

madetech-jeffpinkham replied:
Yep, that's a good call we need to strip that out

zuren commented on LBHTenancyAPI/Controllers/ArrearsActionDiaryController.cs:
```diff
@@ -0,0 +1,38 @@
+using LBHTenancyAPI.Extensions.Validation;
+using LBHTenancyAPI.UseCases.ArrearsActions;
+
+// For more information on enabling MVC for empty projects, visit https://go.microsoft.com/fwlink/?LinkID=397860
```
Probably unnecessary?

madetech-jeffpinkham replied:
The using statements or comment?

madetech-jeffpinkham replied:
That's autogenerated, I'll remove it.

zuren commented on LBHTenancyAPI/Controllers/TenanciesController.cs:
```diff
@@ -58,6 +58,7 @@ public async Task<IActionResult> Get([FromQuery(Name = "tenancy_refs")] List<str
                 {"tenancies", tenancies}
             };
 
+            
```
Unnecessary change?

madetech-jeffpinkham replied:
Yep, I'll reverse

zuren commented on LBHTenancyAPI/Dockerfile:
```diff
@@ -1,12 +1,13 @@
-RUN dotnet restore -nowarn:msb3202,nu1503
+COPY LBHTenancyAPI/LBHTenancyAPI.csproj LBHTenancyAPI/
+RUN ls
+RUN dotnet restore LBHTenancyAPI/LBHTenancyAPI.csproj -nowarn:msb3202,nu1503
```
Looks like there's a debug "ls" left over in here. Are the changes to this file actually necessary now that the config has been fixed, or was this just experimenting?

zuren commented on LBHTenancyAPI/LBHTenancyAPI.csproj:
```diff
@@ -23,5 +28,36 @@
+      <RelPath>Web References\ArrearsAgreementService</RelPath>
+      <UpdateFromURL>http://lbhuhwwebp01/civica-uh-gateway-dev/webservices/ArrearsAgreementService.svc</UpdateFromURL>
+    </WebReferenceUrl>
+  </ItemGroup>-->
```
Should we just remove this config outright, rather than commenting it out? Looks like a security concern to have service config in here.

madetech-jeffpinkham replied:
Agreed

zuren commented on LBHTenancyAPITest/Connected Services/ArrearsAgreementService/ConnectedService.json:
```diff
@@ -0,0 +1,24 @@
+  "ProviderId": "Microsoft.VisualStudio.ConnectedService.Wcf",
+  "Version": "15.0.20119.1312",
+  "ExtendedData": {
+    "Uri": "http://lbhuhwwebp01/civica-uh-gateway-dev/webservices/ArrearsAgreementService.svc",
```
Same here r.e. security.

madetech-jeffpinkham replied:
Yep agreed

zuren commented on LBHTenancyAPITest/Dockerfile:
```diff
@@ -1,12 +1,12 @@
 COPY LBHTenancyAPI.sln ./
-COPY LBHTenancyAPITest/LBHTenancyAPITest.csproj LBHTenancyAPITest/
-RUN dotnet restore -nowarn:msb3202,nu1503
+COPY LBHTenancyAPITest/LBHTenancyAPITest.csproj LBHTenancyAPITest/
```
Same here - necessary change?

madetech-jeffpinkham replied:
removed

zuren commented on LBHTenancyAPITest/Helpers/Fake.cs:
```diff
@@ -122,5 +123,42 @@ public static ArrearsActionDiaryEntry GenerateActionDiary()
+                    ActionBalance = random.Finance.Amount(),
+                    ActionCode = random.Random.Hash(3),
+                    ActionCategory = random.Random.Hash(10),
+                    //Comment = random.Random.Hash(100)
```
Why is this commented out?

madetech-jeffpinkham replied:
Unsure, pulled in from previous branch, checked and there is no comment field. Removed

zuren commented on LBHTenancyAPITest/Helpers/Fake.cs:
```diff
@@ -122,5 +123,42 @@ public static ArrearsActionDiaryEntry GenerateActionDiary()
+                    ActionCategory = random.Random.Hash(10),
+                    //Comment = random.Random.Hash(100)
+                },
+                ErrorCode = 100,
```
ErrorCode 100?

zuren commented on LBHTenancyAPITest/Test/Gateways/ArrearsActions/UHArrearsDiaryCreationGatewayTest.cs:
```diff
@@ -0,0 +1,98 @@
+                    UserName = "HackneyAPI",
+                    UserPassword = "Hackney1"
+                },
+                SourceSystem = "HackneyAPI"
```
Are these real credentials?

madetech-jeffpinkham replied:
Possibly, I'll remove now

zuren commented:
Should probably rebase these commits down to something more manageable/meaningful. Loads of noise from the previous commit history.

StevenLeighton21 commented on LBHTenancyAPI/Extensions/Validation/CreateArrearsActionDiaryExtensions.cs:
```diff
@@ -0,0 +1,34 @@
+        {
+            RuleFor(x => x).NotNull();
+            RuleFor(x => x.ArrearsAction).NotNull();
+            RuleFor(x => x.ArrearsAction.TenancyAgreementRef).NotEmpty().WithMessage("Please specify a first name");
```
I probably wouldn't consider it a priority, but we could add a useful message to each field in that case

StevenLeighton21 commented on LBHTenancyAPI/Gateways/ArrearsActionDiaryGateway.cs:
```diff
@@ -0,0 +1,31 @@
+
+namespace LBHTenancyAPI.Gateways
+{
+    /// <summary>
```
I'd probably omit these unless they're contributing to some doc generation we're using. We aren'y currently doing any generation so would remove them for consistency.

madetech-jeffpinkham replied:
I try to comment where ever possible, especially as a contractor, my time on the project is limited and eventually someone will have to make sense of my work.

StevenLeighton21 commented on LBHTenancyAPITest/Test/Services/ArrearsServiceRequestBuilderTests.cs:
```diff
@@ -0,0 +1,57 @@
+        }
+
+        [Fact]
+        public void WhenGivenActionDiaryDetails_BuildValidArrearsActionDiaryRequest()
```
I think this test would benefit from triangulation, either by randomising the input or by providing a second example test with different input

madetech-jeffpinkham replied:
Ok yep, I'll try and do that

StevenLeighton21 commented on LBHTenancyAPITest/Test/Services/ArrearsServiceRequestBuilderTests.cs:
```diff
@@ -0,0 +1,57 @@
+        [Fact]
+        public void Return_A_Built_Request_Object()
+        {
+            //arrange
```
Suggest removing these comments. Personal preference :)

madetech-jeffpinkham replied:
This will hopefully serve to help the team build good practise in their approach to TDD

StevenLeighton21 commented on LBHTenancyAPITest/Test/Gateways/ArrearsActions/UHArrearsDiaryCreationGatewayTest.cs:
```diff
@@ -0,0 +1,98 @@
+            //act
+            var response = await classUnderTest.CreateActionDiaryEntryAsync(request);
+            //assert
+            Assert.Equal(response.ArrearsAction.TenancyAgreementRef, request.ArrearsAction.TenancyAgreementRef);
```
Suggest these tests could assert against each field in the request, and triangulate the test with a second example/randomised input

madetech-jeffpinkham replied:
Yep, I'll try to do that.

StevenLeighton21 commented on LBHTenancyAPI/Controllers/ArrearsActionDiaryController.cs:
```diff
@@ -0,0 +1,34 @@
+        }
+
+        [HttpPost]
+        public async Task<IActionResult> Post([FromBody][Required] ArrearsActionCreateRequest request)
```
Sorry if this is an obvious question, I don't quite understand how this endpoint receives the info/creates the ArrearsActionCreatRequest. How does an outside consumer use this API?

madetech-jeffpinkham replied:
The swagger docs should help out in this regard. /swagger. But essentially you pass a json object in as a body in a HTTP POST request, with an ContentType of "application/json"

madetech-jeffpinkham commented:
@StevenLeighton21 and @zuren I've actioned the test suggestions, and I'll defer the validation piece to another PR as that's another piece. For the moment I'll hand hold with any API requests until I get that in.

zuren commented on LBHTenancyAPI/Controllers/ArrearsActionDiaryController.cs:
```diff
@@ -0,0 +1,34 @@
+namespace LBHTenancyAPI.Controllers
+{
+    [Produces("application/json")]
+    [Route("api/v1/tenancies/arrearsactiondiary/")]
```
Suggest we change to `/api/v1/tenancies/arrears-action-diary` to be more consistent.

Question around whether we should stick to the "arrears action diary" term, or call this something like `/api/v1/tenancies/arrears_event` ?

madetech-jeffpinkham replied:
Yeah I've been toying with the naming of it the official web service method is called CreateArrearsActionAsync

madetech-jeffpinkham replied:
Arrears actions Controller?

madetech-jeffpinkham replied:
made it /api/v1/tenancies/arrears-action-diary

madetech-jeffpinkham commented:
Ok sounds good :)
